eng/pipelines/coreclr/templates/build-job.yml

```diff
@@ -152,21 +152,105 @@ jobs:
       - name: clrRuntimePortableBuildArg
         value: '-portablebuild=false'
 
+    - name: ConfigurationName
+      value: ${{ format('CoreCLR {0} {1}{2} {3} {4} {5} {6}', parameters.runtimeVariant, parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.buildConfig, parameters.pgoType, parameters.compilerName) }}
+
+    - name: CommitsOutputFile
+      value: '$(Build.SourcesDirectory)/coreclr.commit_hash'
+
+    - name: GetCommitScript
+      value: '$(Build.SourcesDirectory)/src/coreclr/scripts/get_commit_hash.py'
+
+    # Currently only support pipeline caching for PRs and manual triggered runs
+    - ${{ if or(in(variables['Build.Reason'], 'PullRequest'), in(variables['Build.Reason'], 'Manual')) }}:
+      - name: UsePipelineCache
+        value: true
+
+    # Otherwise set all the cache-specific variables to 'false' which will ensure to build product.
+    - ${{ if and(notin(variables['Build.Reason'], 'PullRequest'), notin(variables['Build.Reason'], 'Manual')) }}:
+      - name: UsePipelineCache
+        value: false
+      - name: BUILD_CACHE_RESTORED
+        value: false
+      - name: DAC_BUILD_CACHE_RESTORED
+        value: false
+      - name: BUILD_LOGS_RESTORED
+        value: false
+
+    - ${{ if eq(parameters.osGroup, 'windows') }}:
+      - name: PythonScript
+        value: 'py -3'
+      - name: PipScript
+        value: 'py -3 -m pip'
+
+    - ${{ if ne(parameters.osGroup, 'windows') }}:
+      - name: PythonScript
+        value: 'python3'
+      - name: PipScript
+        value: 'pip3'
+
     - ${{ parameters.variables }}
 
     steps:
 
+    # Get the commit hashes of interesting folder for caching
+    - script: |
+        $(PythonScript) $(GetCommitScript) -configuration "$(ConfigurationName)" -run_type coreclr -output $(CommitsOutputFile)
+      displayName: Record commit hashes of coreclr
+      env:
+        SYSTEM_ACCESSTOKEN: $(System.AccessToken)
+        REPO_URI: $(Build.Repository.Uri)
+      condition: eq(variables.UsePipelineCache, 'true')
+
+    # Cache tasks
+    - task: Cache@2
+      displayName: Cache Build $(ConfigurationName)
+      inputs:
+        key: '"Build" | $(CommitsOutputFile)'
```

Comment:
this should probably also include at least:
- Build machine OS (e.g. we can build the WASM target on both Windows or Linux)
- Arbitrary cache version key, e.g. `v1`, so we can bump it to invalidate the chache (see https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops#q-can-i-clear-a-cache)
- Path to source directory (not sure if this is already implicitly handled by get_commit_hash.py)

I do share Jan's concern that it's hard to capture 100% of all the paths and environment settings that can affect whether a cache can be used.

Reply:
> this should probably also include at least:
> 
> * Build machine OS (e.g. we can build the WASM target on both Windows or Linux)

That should be fairly easy to include. Was not aware about that for WASM.

> * Arbitrary cache version key, e.g. `v1`, so we can bump it to invalidate the chache (see https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops#q-can-i-clear-a-cache)

Good point.

> * Path to source directory (not sure if this is already implicitly handled by get_commit_hash.py)

That is handled by `get_commit_hash.py`.

> I do share Jan's concern that it's hard to capture 100% of all the paths and environment settings that can affect whether a cache can be used.

Sure, the way I understand from the existing comments is there is a consensus in having this just for test artifacts and then experiment if caching and restoring `obj` folder followed by an incremental build would help. If it works, then we can have a job that continuously build and caches the `obj` folder of newer commits and all the PR builds take advantage of them and improving build-time. As @safern pointed, we can have some script to download the `obj` folder depending on the commit you are at, to make inner loop development faster. I tried experimenting the "reusing `obj` folder" on my windows machine by cloning runtime repo in a fresh location and copying the pre-built `obj` folder from another location into it. So far, I see that `obj` folder contains lot of files that has hardcoded path of the original repo location. I tried replacing the paths with new locations, but there were still some places that referenced old repo location and hence when I do a build on new repo location, it was still building it for old location (at least that's what I understood from the console output). I am not too familiar with how it works, or what needs to be updated in `obj` folder for it to get reused. If someone can try this out and confirm the necessary steps, I am happy to take it forward from there.